### PR TITLE
Move `--beta-features` flag to common options; Update `sam --info` to show available beta feature flag env vars

### PR DESCRIPTION
**Pull request by hawflau**
#### Which issue(s) does this change fix?

#4957 

#### Why is this change necessary?
- `--beta-features` is missed in `sam sync`. I think `--beta-features` should be a common flag that's available to all commands, like `--debug`.

#### How does it address the issue?
- Move `--beta-features` to common options.
- Refactor help text formatters for commands "init", "build", "sync" and "deploy" (help text formatter is not implemented for other commands yet)
- Also update `sam --info` to provide info about beta feature flag env var names:

image

#### What side effects does this change have?
No

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/commands/build/core/options.py [hunk 1]
```diff
@@ -76,6 +71,5 @@
             ),
         ],
     },
-    "Beta Options": {"option_names": {opt: {"rank": idx} for idx, opt in enumerate(BETA_OPTIONS)}},
-    "Other Options": {"option_names": {opt: {"rank": idx} for idx, opt in enumerate(OTHER_OPTIONS)}},
 }
+add_common_options_info(OPTIONS_INFO)
```
Comment (sriram-mv) [inline samcli/commands/build/core/options.py:75]: How does this play with imports? there are places where OPTIONS_INFO is imported, does OPTIONS_INFO get mutated prior to import?

File: samcli/cli/main.py [hunk 2]
```diff
@@ -54,6 +56,7 @@ def print_info(ctx, param, value):
         "version": __version__,
         "system": gather_system_info(),
         "additional_dependencies": gather_additional_dependencies_info(),
+        "available_beta_feature_env_vars": get_all_experimental_env_vars(),
```
Comment (sriram-mv) [inline samcli/cli/main.py:59]: Do customers know how to toggle these env vars? Is that documented anywhere?

File: samcli/commands/build/core/options.py (hunk 1 above)
Comment (hawflau) [reply samcli/commands/build/core/options.py:75]: When anything of the module or the module itself is imported, the entire file got executed, meaning that `add_common_options_info(OPTIONS_INFO)` is also executed when `samcli.commands.build.core.options` is imported.

I did a simple experiment to verify this behavior:
```
.
├── __init__.py
├── const.py
└── main.py
```

const.py:
```python
MY_DICT = {}

MY_DICT["foo"] = "bar"
```

main.py
```python
from const import MY_DICT

print(MY_DICT)
```

result:
```
» python main.py
{'foo': 'bar'}
```

File: samcli/cli/main.py (hunk 2 above)
Comment (hawflau) [reply samcli/cli/main.py:59]: Good catch.

We mentioned about `SAM_CLI_BETA_RUST_CARGO_LAMBDA` in the [Rust Builder page](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/building-rust.html#building-rust-prerequisites). But I couldn't find other env vars mentioned in anywhere else.

I believe we should have a doc talking about beta features or about all SAM CLI env vars.

File: samcli/cli/core/command.py [hunk 3]
```diff
@@ -40,7 +40,9 @@ def _format_options(
         params: List[Parameter],
         formatter: RootCommandHelpTextFormatter,
         formatting_options: Dict[str, Dict],
+        write_rd_overrides: Optional[Dict[str, Any]] = None,
     ):
+        write_rd_overrides = write_rd_overrides if write_rd_overrides else {}
```
Comment (mndeveci) [inline samcli/cli/core/command.py:45]: nit: simpler version;
```suggestion
        write_rd_overrides = write_rd_overrides or {}
```

**Review by mndeveci — APPROVED** [review]
LGTM , one small comment

**Pull request merged**